osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using Amazon;
+using Amazon.Runtime;
+using Amazon.S3;
+using Amazon.S3.Model;
+using osu.Framework.Extensions;
+using osu.Game.IO.Archives;
+using osu.Server.BeatmapSubmission.Configuration;
+using osu.Server.BeatmapSubmission.Models;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Zip;
+
+namespace osu.Server.BeatmapSubmission.Services
+{
+    public class S3BeatmapStorage : IBeatmapStorage
+    {
+        private const string osz_directory = "osz";
+        private const string versioned_file_directory = "beatmap_files";
+
+        private readonly AmazonS3Client client;
+
+        public S3BeatmapStorage()
+        {
+            client = new AmazonS3Client(
+                new BasicAWSCredentials(AppSettings.S3AccessKey, AppSettings.S3SecretKey),
+                new AmazonS3Config
+                {
+                    CacheHttpClient = true,
+                    HttpClientCacheSize = 32,
+                    RegionEndpoint = RegionEndpoint.USWest1,
+                    UseHttp = true,
+                    ForcePathStyle = true,
+                    RetryMode = RequestRetryMode.Legacy,
+                    MaxErrorRetry = 5,
+                    Timeout = TimeSpan.FromMinutes(1),
+                });
```

Comment:
Note that [contrary to `osu-server-spectator`](https://github.com/ppy/osu-server-spectator/blob/89f47685d34ff42031847ac475ec5d9293cb7eed/osu.Server.Spectator/S3.cs) (and more?) this is created in the ctor. This service also has a singleton lifetime. This means that this object is _shared_.

I suspect that `osu-server-spectator` is very wrong in instantiating `AmazonS3Client` on every call. The reason why is that `HttpClientCacheSize` is supposed to specify how many `HttpClient` instances the amazon client can use for requests. Which would mean that every instantiation of `AmazonS3Client` spawns 32 `HttpClient`s. [The guidance on `HttpClient` usage](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines) is that it is supposed to be *shared*, as in you're not supposed to create one every call.

(All of that would be nullified if it turned out `AmazonS3Client` is doing some black magic hackery with `static` to ensure the http client cache is created once ever, but on a quick source inspection I do not believe that to be true.)

Reply:
Also there's an argument here that rather than doing its own `HttpClient` pooling this should be feeding into the ASP.NET-provided `HttpClient` DI factory machinery. I would be OK with making it do that, but am lacking context as to why this `HttpClientCacheSize` was being set to what it was (probably originates from 5 projects back and kept getting copy-pasted around as convenient).

Reply:
Sounds very suspicious. I do recall that this should be a singleton by design, so I think we probably want to fix that in `server-spectator` as well.